src/libsyntax/config.rs

```diff
@@ -49,8 +56,25 @@ impl<F> fold::Folder for Context<F> where F: FnMut(&[ast::Attribute]) -> bool {
         fold_item_underscore(self, item)
     }
     fn fold_expr(&mut self, expr: P<ast::Expr>) -> P<ast::Expr> {
+        // If an expr is valid to cfg away it will have been removed by the
+        // outer stmt or expression folder before descending in here.
+        // Anything else is always required, and thus has to error out
+        // in case of a cfg attr.
+        //
+        // NB: This is intentionally not part of the fold_expr() function
+        //     in order for fold_opt_expr() to be able to avoid this check
```

Comment:
when I first read this comment, I interpreted the phrase "to be able to avoid this check" as some sort of efficiency concern (which admittedly seemed a little strange in this context).

But now I think I understand: `fold_opt_expr` *must* skip the check, because the check does not apply to optional expression contexts.  Do I read that right?

(I'm not sure I can think of a better phrasing off-hand.  It could be just me that strongly associates the phrase "avoid a check" with "micro-optimization"... :)